integration-test/src/test/scala/com/cloudera/livy/test/BatchIT.scala

```diff
@@ -78,12 +94,79 @@ class BatchIT extends BaseIntegrationTestSuite {
     assert(result.state === SessionState.Success().toString)
   }
 
-  private def newOutputPath(): String = {
-    cluster.hdfsScratchDir().toString() + "/" + UUID.randomUUID().toString()
+  test("deleting a session should kill YARN app") {
+    assume(testLibPath != null, "Test lib not uploaded.")
+    val output = newOutputPath
+    val batchId = runSpark(classOf[SimpleSparkApp], List(output, "false"), waitForExit = false).id
+
+    dumpLogOnFailure(batchId, cleanup = false) {
+      val appId = waitBatchForRunning(batchId)
+
+      // Delete the session then verify the YARN app state is KILLED.
+      deleteBatch(batchId)
+      assert(cluster.yarnClient.getApplicationReport(appId)
+        .getFinalApplicationStatus == FinalApplicationStatus.KILLED,
+        "YARN app should be killed.")
+    }
+  }
+
+  test("killing YARN app should change batch state to dead") {
+    assume(testLibPath != null, "Test lib not uploaded.")
+    val output = newOutputPath
+    val batchId = runSpark(classOf[SimpleSparkApp], List(output, "false"), waitForExit = false).id
+
+    dumpLogOnFailure(batchId) {
+      val appId = waitBatchForRunning(batchId)
+
+      // Kill the YARN app and check batch state should be KILLED.
+      cluster.yarnClient.killApplication(appId)
+      eventually {
+        assert(cluster.yarnClient.getApplicationReport(appId)
+          .getFinalApplicationStatus == FinalApplicationStatus.KILLED,
+          "YARN app should be killed.")
+        val batchInfo = getBatchSessionInfo(batchId)
+        assert(batchInfo.state == SessionState.Dead().toString, "Batch state should be dead.")
+        assert(batchInfo.log.contains("Application killed by user."),
+          "Batch log doesn't contain yarn final diagnostics.")
+      }
+    }
+  }
+
+  private def dumpLogOnFailure[T](batchId: Int, cleanup: Boolean = true)(f: => T): T = {
+    try {
+      f
+    } catch {
+      case e: Throwable =>
+        allCatch {
+          info(s"Session log: ${getBatchSessionInfo(batchId).log}")
+          info(s"YARN log: ${getYarnLog(batchId)}")
+        }
+        throw e
+    } finally {
+      if (cleanup) {
+        deleteBatch(batchId)
+      }
+    }
+  }
+
+  private def getYarnLog(batchId: Int): String = {
+    allCatch.opt {
+      val appId = getBatchSessionInfo(batchId).appId
+      assert(appId != null, "appId shouldn't be null")
+
+      val appReport = cluster.yarnClient.getApplicationReport(ConverterUtils.toApplicationId(appId))
+      assert(appReport != null, "appReport shouldn't be null")
+
+      appReport.getDiagnostics
+    } getOrElse ""
+  }
+
+  private def newOutputPath: String = {
```

Comment:
This is more of a style convention; but if a method is not a simple property getter, it's good practice to use `()` since it makes it a little clearer that the method actually does things.